README.md
# What is it?

IGHTMLQuery is a lightweight XML/HTML parser for iOS, built on top of libxml. It is inspired by jQuery and nokogiri. Consider following snippets:

```objective-c
IGXMLDocument* node = [[IGXMLDocument alloc] initWithXMLString:catelogXml error:nil];
NSString* title = [[[node queryWithXPath:@"//cd/title"] firstObject] text];
[[node queryWithXPath:@"//title"] enumerateNodesUsingBlock:^(IGXMLNode *title, NSUInteger idx, BOOL *stop) {
    NSLog(@"title = %@", title.text);
}];

// or use CSS Selector
[[node queryWithCSS:@"title"] enumerateNodesUsingBlock:^(IGXMLNode *title, NSUInteger idx, BOOL *stop) {
    NSLog(@"title = %@", title.text);
}];

// quick manipulation
[[node queryWithXPath:@"//title"] appendWithXMLString:@"<message>Hi!</message>"];
```

## Features

- Use XPath and CSS Selector for document searching.
- jQuery style chainable syntax.
- XML traversal and manipulation.

## Installation

IGHTMLQuery is available through [CocoaPods](http://cocoapods.org/), to install it simply add the following line to your Podfile:

```ruby
pod "IGHTMLQuery", "~> 0.8.4"
```

Alternatively:

1. Add all the source files in ```Classes``` to your Xcoe project
2. In "Build Phases" > "Link Binary With Libraries, add ```libxml2```.
3. In "Build Setting", find "Header Search Paths" and add "$(SDK_DIR)/usr/include/libxml2"

## Usage

### Import header

For each files using IGHTMLQuery, import ```IGHTMLQuery.h```:

```
#import 'IGHTMLQuery.h'
```

### Create a document

Create a XML document:

```objective-c
IGXMLDocument* node = [[IGXMLDocument alloc] initWithXMLString:xml error:nil];
```

Create a HTML document:

```objective-c
IGHTMLDocument* node = [[IGHTMLDocument alloc] initWithHTMLString:html error:nil];
```

### Traversal

Use ```parent```, ```nextSibling```, ```previousSibling```, ```children``` and ```firstChild``` to traverse the document.

### Query using XPath or CSS Selector

You can query the document or any node with ```queryWithXPath:``` or ```queryWithCSS:``` methods. They will always return a ```IGXMLNodeSet``` object, which is a set like object that you can chain query and operations.

```objective-c
IGXMLNodeSet* contents = [doc queryWithXPath:@"//div[@class='content']"];
[contents enumerateNodesUsingBlock:^(IGXMLNode* content, NSUInteger idx, BOOL *stop){
    NSLog(@"%@", content.xml);
}];

// use a @try/@catch block for queryWithCSS, as it can throw an exception if 
// the CSS Selector cannot be converted to XPath.
@try {
  contents = [doc queryWithCSS:@"div.content"];
  [contents enumerateNodesUsingBlock:^(IGXMLNode* content, NSUInteger idx, BOOL *stop){
      NSLog(@"%@", content.xml);
  }];
} @catch(NSException * e) {
  // handle error
}
```

### Document Manipulation

You can change the document using methods in ```IGXMLNodeManipulation``` protocol.

```objective-c
@protocol IGXMLNodeManipulation <NSObject>

-(instancetype) appendWithNode:(IGXMLNode*)child;

-(instancetype) prependWithNode:(IGXMLNode*)child;

-(instancetype) addChildWithNode:(IGXMLNode*)child;

-(instancetype) addNextSiblingWithNode:(IGXMLNode*)child;

-(instancetype) addPreviousSiblingWithNode:(IGXMLNode*)child;

-(void) empty;

-(void) remove;

@end

```

## JavaScript/Ruby support

All classes in IGHTMLQuery supports JavaScriptCore exports. Additionally there
are Ruby wrappers to used with [JavaScriptCoreOpalAdditions](https://github.com/siuying/JavaScriptCoreOpalAdditions), which allow you to manipulate DOM with Ruby in Objective-C like this ...

```objective-c
#import "JSContext+IGHTMLQueryRubyAdditions.h"
#import "JSContext+OpalAdditions.h"

// load IGHTMLQuery ruby wrapper classes
[context configureIGHTMLQuery];

// create a lambda that evalulate script on the fly
JSValue* instanceEval = [context evaluateRuby:@"lambda { |doc, script| XMLNode.new(doc).instance_eval(&eval(\"lambda { #{script} }\")) }"];

// a simple script that find the title of first cd have a price less than 9.0
JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').find {|node| node.xpath('./price').text.to_f < 9.0 }.xpath('./title').text"]];

// convert the result to string
NSString* title = [node toString];
XCTAssertEqualObjects((@"Greatest Hits"), title, @"title should be Greatest Hits");
```

To use IGHTMLQuery with Ruby support, add following line to your Podfile:

```ruby
pod "IGHTMLQuery/Ruby"
```

See [Test Cases](https://github.com/siuying/IGHTMLQuery/blob/master/IGHTMLQueryTests/IGRubyTests.m) for more detail.

## Breaking Changes

### 0.7.2

In previous version, the method ``[[IGHTMLDocument alloc] initWithHTMLString:]`` will create html element without implied HTML tag. (HTML_PARSE_NOIMPLIED option in libxml). Since 0.7.2, HTML_PARSE_NOIMPLIED will no longer be the default.

If you want to maintain the old behavior, check the ``[[IGHTMLDocument alloc] initWithHTMLFragmentString:]`` method.

## License

MIT License.


CHANGELOG.md
Changelog
=========

0.8.4
=====

* Use a per thread cached instance of CSSSelectorConverter.

0.8.3
=====

* Use a shared instance of CSSSelectorConverter.

0.8.2
=====

* Fix a bug cause nested CSS query not work as expect.

0.8.1 (13/3/2014)
-----------------

* Added: ``[node innerHtml]`` and ``[node html]``. Like ``[node innerXml]`` and ``[node xml]``, they return the markup of the node. If you are working on HTML document, you should use the html version of the method, as they handle self-closing tag differently.  

0.8.0 (7/3/2014)
-----------------
* Added: [node query:] and [nodeSet query:] which automatically select query using XPath or CSS Selector.
* Fixes bug for IGXMLNode equality.
* [nodeSet enumerateNodesUsingBlock:] now enumerate the order of the set instead of reversed.

0.7.2 (25/2/2014)
-----------------
* Remove the HTML_PARSE_NOIMPLIED option in HTML parser.

0.7.1 (28/1/2014)
-----------------
* Use CSSSelectorConverter 1.1.0, fixes critical selector match issue.

0.7.0 (23/1/2014)
-----------------
+ Introduce [node queryWithCSS:], query node using CSS Selector Level 3 (powered by [CSSSelectorConverter](https://github.com/siuying/CSSSelectorConverter))
- [Ruby] Fixes a bug wrapping return values.

0.6.6 (5/1/2014)
----------------
- [JavaScript] Fixed a bug using IGHTMLDocument

0.6.5 (4/1/2014)
----------------
+ [JavaScript] add HTMLDoc, allow creation of document with HTML on JavaScript side.
+ Add IGHTMLQuery::HTTP.get method, that a simple wrapper to perform a HTTP request.

IGHTMLQueryTests/IGRubyTests.m
//
//  IGRubyTests.m
//  IGHTMLQuery
//
//  Created by Francis Chong on 12/26/13.
//  Copyright (c) 2013 Ignition Soft. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <JavaScriptCore/JavaScriptCore.h>
#import <JavaScriptCoreOpalAdditions/OpalCore.h>

#import "JSContext+OpalAdditions.h"
#import "JSContext+IGHTMLQueryRubyAdditions.h"
#import "IGHTMLQuery.h"
#import "IGHTMLQueryJavaScriptExport.h"

@interface IGRubyTests : XCTestCase
{
    IGXMLDocument* doc;
    JSContext* context;
    JSValue* instanceEval;
}
@end

@implementation IGRubyTests

- (void)setUp
{
    [super setUp];
    
    NSString* content = [[NSString alloc] initWithContentsOfFile:[[NSBundle bundleForClass:[self class]] pathForResource:@"catalog" ofType:@"xml"] encoding:NSUTF8StringEncoding error:nil];
    doc = [[IGXMLDocument alloc] initWithXMLString:content error:nil];
    context = [[JSContext alloc] init];
    [context configureIGHTMLQuery];

    // setup context
    context[@"doc"] = doc;
    
    // instance_eval on a doc
    instanceEval = [context evaluateRuby:@"lambda { |doc, script| XMLNode.new(doc).instance_eval(&eval(\"lambda { #{script} }\")) }"];
}

- (void)testNodeCore
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.tag"]];
    XCTAssertEqualObjects(@"catalog", [node toString], @"should be catalog");
}

- (void)testNodeAttribute
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.children.first['country'] = 'Hong Kong'; self.children.first['country'] "]];\
    XCTAssertEqualObjects(@"Hong Kong", [node toString], @"should be Hong Kong");

    node = [instanceEval callWithArguments:@[doc, @"self.children.first['abc'].nil?"]];\
    XCTAssertEqualObjects(@YES, [node toNumber], @"non existing attribute should return nil");
}

- (void)testNodeQuery
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd/price').first.text.to_f"]];
    XCTAssertEqualWithAccuracy(10.9, [[node toNumber] doubleValue], 0.001, @"should find first price");
}

- (void)testNodeCSSQuery
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.css('cd > price').first.text.to_f"]];
    XCTAssertEqualWithAccuracy(10.9, [[node toNumber] doubleValue], 0.001, @"should find first price");
    
    node = [instanceEval callWithArguments:@[doc, @"self.css('cd[country=\"UK\"] price').text.to_f"]];
    XCTAssertEqualWithAccuracy(9.9, [[node toNumber] doubleValue], 0.001, @"should find first price");
}

- (void)testNodeAttributes
{
    JSValue* attributes = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').first.attributes"]];
    XCTAssertEqual(1U, [[attributes toArray] count], @"should have attributes");
    XCTAssertEqualObjects(@"country", [attributes[0] toString], @"should attribute country");
}

- (void)testNodeManipulationRemove
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').each {|n| n.remove }; self.xpath('//cd').nodes "]];
    XCTAssertEqual(0U, [[node toArray] count], @"should remove cds");
}

- (void)testNodeManipulationEmpty
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.empty; self.xml"]];
    XCTAssertEqualObjects(@"<catalog/>", [node toString], @"should empty xml");
}

- (void)testNodeToNative
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.to_n"]];
    IGXMLNode* nodeNative = [node toObject];
    XCTAssertNotNil(nodeNative, @"should be a node");
    XCTAssertTrue([nodeNative isKindOfClass:[IGXMLNode class]], @"should be a node");
    XCTAssertEqualObjects(@"catalog", nodeNative.tag, @"should be a catalog");
}

#pragma mark - Node Set

- (void)testNodeSet
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"first = self.children.all.first; first.tag"]];
    XCTAssertEqualObjects(@"cd", [node toString], @"should be cd");
    
}

- (void)testNodeSetManipulation
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').remove; self.xpath('//cd').nodes "]];
    XCTAssertEqual(0U, [[node toArray] count], @"should remove cds");
}

- (void)testNodeSetEnumerable
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').find {|node| node.xpath('./price').text.to_f < 9.0 }.xpath('./title').text"]];
    NSString* title = [node toString];
    XCTAssertEqualObjects((@"Greatest Hits"), title, @"title should be Greatest Hits");
}

- (void)testNodeSetToNative
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').to_n"]];
    IGXMLNodeSet* titleNodeSet = [node toObject];
    XCTAssertNotNil(titleNodeSet, @"should be a node set");
    XCTAssertEqualObjects(NSStringFromClass([titleNodeSet class]), NSStringFromClass([IGXMLNodeSet class]), @"should be a NodeSet");
    XCTAssertEqual(3U, [titleNodeSet count], @"should have 3 nodes in set");
}

#pragma mark - Traversal

- (void)testParent
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').first.parent.tag"]];
    NSString* tag = [node toString];
    XCTAssertEqualObjects(tag, @"catalog", @"should be catalog");
}

- (void)testNoParent
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//catalog').first.parent.parent.nil?"]];
    XCTAssertEqualObjects(@YES, [node toNumber], @"non existing parent return nil");
}

- (void)testChildrenAndFirstChild
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').first.children.first.tag"]];
    XCTAssertEqualObjects(@"title", [node toString], @"should be title");
    
    node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').first.first_child.tag"]];
    XCTAssertEqualObjects(@"title", [node toString], @"should be title");
}

- (void)testNoChildren
{
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').first.first_child.nil?"]];
    XCTAssertEqualObjects(@NO, [node toNumber], @"should not be nil");

    node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd').first.first_child.first_child.nil?"]];
    XCTAssertEqualObjects(@YES, [node toNumber], @"should be nil");
}

- (void)testNextAndPreviousSibling
{
    IGXMLNode* ukCd = [[doc queryWithXPath:@"cd[@country='UK']"] firstObject];
    
    XCTAssertEqualObjects([ukCd.previousSibling queryWithXPath:@"price"].firstObject.text, @"10.90");
    XCTAssertEqualObjects([ukCd.nextSibling queryWithXPath:@"price"].firstObject.text, @"8.90");
    
    JSValue* node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd[@country=\"UK\"]').first.previous_sibling.xpath('price').first.text"]];
    XCTAssertEqualObjects(@"10.90", [node toString], @"should not be nil");

    node = [instanceEval callWithArguments:@[doc, @"self.xpath('//cd[@country=\"UK\"]').first.next_sibling.xpath('price').first.text"]];
    XCTAssertEqualObjects(@"8.90", [node toString], @"should not be nil");
}

#pragma mark - Others

- (void) testHTMLDoc {
    JSValue* h1 = [context evaluateRuby:@"HTMLDoc.new('<html><body><h1>Header</h1><div id=\"content\"><span>Hello</span></div></body></html>').xpath('//h1').first.text"];
    XCTAssertEqualObjects(@"Header", [h1 toString], @"should return a doc based on input and can be queried");
    
    h1 = [context evaluateRuby:@"HTMLDoc.new('<html><body><h1>Header</h1><div id=\"content\"><span>Hello</span></div></body></html>').css('h1').first.text"];
    XCTAssertEqualObjects(@"Header", [h1 toString], @"should return a doc based on input and can be queried");
    
}

-(void) testHTTPGet {
    JSValue* oldGet = context[@"HTTPGet"];
    context[@"HTTPGet"] = ^NSString*(NSString* url){ return @"HELLO"; };
    NSString* wiki = [[context evaluateRuby:@"IGHTMLQuery::HTTP.get('http://www.wikipedia.org')"] toString];
    XCTAssertEqualObjects(wiki, @"HELLO", @"should return with HTTPGet method");
    context[@"HTTPGet"] = oldGet;
}

@end
